Support for the Xiaomi Induction Cooker

basveeling opened the pull request:
This is a first attempt at supporting the Xiaomi Induction Cooker models (#491), which include models
`chunmi.ihcooker.eg1 chunmi.ihcooker.exp1 chunmi.ihcooker.chefnic chunmi.ihcooker.hk1 chunmi.ihcooker.korea1 chunmi.ihcooker.tw1 chunmi.ihcooker.v1`

Turns out the device is a bit more advanced than the app makes it appear. It supports 16-step recipes with each step having specific power, temperature, timer and next-phase settings. This enables building 'smart' recipes for e.g. automated rice cooking, kettle, pressure cooking, etc. 

Supported features:
- [x] Stopping device
- [x] Starting a profile with and without confirmation
- [x] Status of cooker including temperature, power/fire output, recipe phase.
- [x] Saving a recipe to the menu.
- [x] Making custom recipes with `CookProfile`.
- [x] Update recipe settings while running.
- [x] Factory reset.
- [ ] Interpreting existing recipes.

To do:
- [ ] Build successfully; currently stuck on an issue with poetry/tox `AttributeError: type object 'Callable' has no attribute '_abc_registry'` <- I could use some advice here :-).
- [ ] Run some more real-world tests.
- [ ] Verify with beta testers.
- [ ] Verify that other models are correctly supported.
- [ ] Figure out some of the remaining mystery bits & bytes in the recipe and status fields.
- [ ] Write more tests on integration.

Shout-out to @aquarat, @EUA and @hossF for initial analysis and @esgie for the plugin APK 👍 

Here's a read-out of a rice-cooker program.
![image](https://user-images.githubusercontent.com/536975/95659712-3839b900-0b23-11eb-9394-3ac4bf4b6dc0.png)

coolibry commented:
Hi basveeling,

thanks for this works.

for Build successfully; currently stuck on an issue with poetry/tox AttributeError: type object 'Callable' has no attribute '_abc_registry' <- I could use some advice here :-) :
maybe it's on line 621 of ihcooker.py, change menu to recip_name :
"Menu: {result.recipe_name}\n"

in my side build works.

didjeru commented:
Any news on this issue?

guancio commented:
I'm willing to test it. It is working on my induction cooker.

rytilahti commented on miio/ihcooker.py:
```diff
@@ -0,0 +1,746 @@
+
+def calcrc(data, n):
+    """
+     public static int calcrc(byte[] bArr, int i) {
+        int i2 = 0;
+        int i3 = 0;
+        while (true) {
+            i--;
+            if (i < 0) {
+                return 65535 & i2;
+            }
+            int i4 = i3 + 1;
+            int i5 = i2 ^ (bArr[i3] << 8);
+            for (int i6 = 0; i6 < 8; i6++) {
+                i5 = (32768 & i5) != 0 ? (i5 << 1) ^ 4129 : i5 << 1;
+            }
+            i2 = i5;
+            i3 = i4;
+        }
+    }
```
This is unnecessary comment. Do you happen to know what sort of checksum function this is, if it's a common one this should simply be replaced with an invocation to an existing implementation to make the code more readable.

basveeling replied:
This turned out to be CRC-CCITT, thanks to @nekromant  https://github.com/syssi/xiaomi_cooker/issues/15#issuecomment-848635066 .

rytilahti commented on miio/ihcooker.py:
```diff
@@ -0,0 +1,746 @@
+        else:
+            self.data = bytearray([0] * 179)
+            # Initialize recipe phases.
+            if self.is_v1:
+                offset = 38
+            else:
+                offset = 48
+
+            for i in range(offset, 16 * 8, 8):
+                self.data[i + 0] = 0
+                self.data[i + 1] = 0
+                self.data[i + 2] = 0
+                self.data[i + 3] = 249
+                self.data[i + 4] = 229
+                self.data[i + 5] = 0
+                self.data[i + 6] = 20
+                self.data[i + 7] = 20
```
If you know (at least parts) of the structure, it'd be great to have this in `construct` structs as used for the miio protocol itself for maintainability, see https://construct.readthedocs.io/en/latest/intro.html

nekromant replied:
I wonder if the format is similar in any way to chunmi.cooker.normalX models. Did you happen to document any of your reverse engineering efforts?

coolibry replied:
from my knowing :
`self.data[i + 0] = 0 # mode: heat_until_temp, heat_until_timer, continuous_output.
self.data[i + 1] = 0 # hours: how long this phase lasts in minutes.
self.data[i + 2] = 0 #mins: how long this phase lasts in minutes.
self.data[i + 3] = 249 #thresh: temperature threshold for moving to next phase, in celcius.
self.data[i + 4] = 229 #temp: target temperature to heat to, in celsius.
self.data[i + 5] = 0 #fire: power output between [0,99]

#set continuous or pulse fire
#same value for 2 = continous heat
#different value = pulse heat, usefull for low heat adjustment
#0 = no heat
self.data[i + 6] = 20 #fire off time
self.data[i + 7] = 20 #fire on time`

nekromant replied:
@coolibry Thanks, will tinker and see if it fits my cooker.

basveeling replied:
@rytilahti : Using Construct here makes sense I think. I propose to leave this open for follow-up pull request though.

rytilahti commented on miio/ihcooker.py:
```diff
@@ -0,0 +1,746 @@
+
+    def set_start_remind(self, value):
+        """Prompt user to start recipe, used with set_menu function."""
+        if self.is_v1:
+            i = 21
+        else:
+            i = 37
+        if value:
+            self.data[i] = self.data[i] | 64
+        else:
+            self.data[i] = self.data[i] & 190
```
Magic values should be constified.

basveeling replied:
Done

rytilahti commented on miio/ihcooker.py:
```diff
@@ -0,0 +1,746 @@
+        else:
+            offset = 48
+        temp_target = 0
+        for phase_i in range(16):
+            o = offset + phase_i * 8
+            if phase_i >= len(phases):
+                self.data[o + 0] = 0
+                self.data[o + 1] = 0  # 128
+                self.data[o + 2] = 0
+                self.data[o + 3] = 229
+                self.data[o + 4] = temp_target
+                self.data[o + 5] = 0
+                self.data[o + 6] = 20
+                self.data[o + 7] = 20
+            else:
```
See the earlier comments.

basveeling replied:
Done!

rytilahti reviewed (changes requested):
Hi and thanks for the PR! Looks like I forgot to submit this review looooong time ago for some reason, apologies for that... I hope that the inlined comments should help restructuring the code a bit, but it's currently just way too much to review piece-by-piece.

The biggest issue I have in its current form is that the code is too hard to read: there are too many magic numbers and direct index accesses (also outside the checksum calculation) that should be easier to define as `contruct.Struct`s for maintainability.

Considering that at least part of the contents of those hex-encoded payloads are already known based on the code, it should not be too hard to convert them to use construct for parsing.

rytilahti commented:
There is also #1018 (which aims to support only one model listed in your description). I think it would be great if you @basveeling and @sschirr could combine your efforts & knowledge on this topic :-)

basveeling commented on miio/ihcooker.py:
```diff
@@ -0,0 +1,746 @@
+                hours = minutes // 60
+                minutes = minutes % 60
+
+                if mode == "heat_until_temp" or mode == "2":
```
Agreed, switched to enum.

basveeling commented:
> There is also #1018 (which aims to support only one model listed in your description). I think it would be great if you @basveeling and @sschirr could combine your efforts & knowledge on this topic :-)

It seems this is a different device class (chunmi.cooker, whereas this PR concerns chunmi.ihcooker) and the multicooker seems closer to the rice cookers in operation.

basveeling commented on miio/ihcooker.py:
```diff
@@ -0,0 +1,746 @@
+                    self.data[o + 2] = 45
+                    self.data[o + 3] = temp_threshold
+                elif mode == "16":
+                    self.data[o + 0] = 16  # not sure
```
See below.

basveeling commented on miio/ihcooker.py:
```diff
@@ -0,0 +1,746 @@
+    Finish = "finish"
+    Offline = "offline"
+    Pause = "pause"
+    Pause_time = "pause_time"
```
Done!

basveeling commented:
@rytilahti Thank you for the review! With a bit of delay on my side, I believe I have made most of the changes requested, outside of moving the bytestring building to Construct. I can take a look at that at a later point, I leave it up to you to decide if that is a blocking issue for this PR. In the mean time I'm happy to address any other issues, looking forward to hearing your thoughts!

basveeling commented:
Hi Teemu / @rytilahti,

I hope all is well. I see there's still a change requested status on this PR, but I think I've addressed all the comments. Just wanted to make sure it's clear this is ready for your review whenever you find the time :-).

Enjoy the holiday period!

rytilahti commented on miio/integrations/ihcooker/ihcooker.py:
```diff
@@ -0,0 +1,565 @@
+    def delete_recipe(self, location):
+        """Delete recipe at location [0,7]"""
+        if location >= 9 or location < 1:
+            raise IHCookerException("location %d must be in [1,8]." % location)
```
`start_fire` is using `ValueError`, maybe consolidate this (and others if necessary) to raise that standard error on invalid values?

basveeling replied:
Done!

rytilahti commented on miio/integrations/ihcooker/ihcooker.py:
```diff
@@ -0,0 +1,565 @@
+
+        :arg
+
+        Please do not use skip_confirmation=True, as this is potentially unsafe.
```
How unsafe this is? Unsafe enough to avoid exposing it? :-)

basveeling replied:
Good question. The device itself has a couple of fail-saves: automatically stops if there's no ferrous metal on the cooktop, and overheating protection if the temp sensor measures above 250c if I'm not mistaken. I can imagine remote start can be a nice feature for certain projects. What's your stance on this?

rytilahti replied:
I think it's fair enough to keep it available, anyone toggling this on should know what they are doing.

rytilahti commented on miio/integrations/ihcooker/ihcooker.py:
```diff
@@ -0,0 +1,565 @@
+    )
+    def start(self, profile: Union[str, c.Container, dict], skip_confirmation=False):
+        """Start cooking a profile.
+
+        :arg
```
```suggestion
```

rytilahti commented on miio/integrations/ihcooker/ihcooker.py:
```diff
@@ -0,0 +1,565 @@
+        """Blue wifi led setting at bottom of device: true if led remains on at idle."""
+        if "set_wifi_led" in self.data:
+            return bool(self.data["set_wifi_led"] == "01")
+        else:
+            return None
```
```suggestion

        return None
```

rytilahti commented on miio/discovery.py:
```diff
@@ -174,6 +183,13 @@
     "chunmi-cooker-normal3": Cooker,
     "chunmi-cooker-normal4": Cooker,
     "chunmi-cooker-normal5": Cooker,
+    MODEL_EXP1: IHCooker,
+    MODEL_FW: IHCooker,
+    MODEL_TW1: IHCooker,
+    MODEL_KOREA1: IHCooker,
+    MODEL_HK1: IHCooker,
+    MODEL_V1: IHCooker,
+    MODEL_EG1: IHCooker,
```
I don't think these are correct for mdns? Either remove these completely for now, or check what format is being used in mdns on your device and just add them here as regular strings.

This information should be moved inside the integration itself at some point, but it'll require some more work. Having these here as strings makes it easy to copy them over if/when that time comes.

basveeling replied:
Removed.

rytilahti replied:
They seem to be still there in the diff :eyes:

rytilahti commented on miio/discovery.py:
```diff
@@ -8,6 +8,16 @@
 import zeroconf
 
+from miio import IHCooker
+from miio.integrations.ihcooker import (
+    MODEL_EG1,
+    MODEL_EXP1,
+    MODEL_FW,
+    MODEL_HK1,
+    MODEL_KOREA1,
+    MODEL_TW1,
+    MODEL_V1,
+)
```
```suggestion
```

rytilahti commented on docs/api/miio.ihcooker.rst:
```diff
@@ -0,0 +1,7 @@
+miio.ihcooker module
+====================
+
+.. automodule:: miio.ihcooker
```
Either remove this file or fix the hierarchy.

basveeling replied:
Removed it.

rytilahti replied:
It's still here :-)

rytilahti reviewed (changes requested):
Hey Bas, sorry, I completely forgot to submit some commentary I had written earlier... Thanks for the reminder! So there are just a couple of minor changes I'd like to see prior merging this, after those are done this is good to go :+1:

basveeling commented on miio/integrations/ihcooker/tests/test_ihcooker.py:
```diff
@@ -0,0 +1,321 @@
+
+    def test_crc(self):
+        recipe = {}
+        self.device.start(recipe)
```
That's right, this was helpful during refactoring the codebase. It might also be helpful to clarify what kind of inputs these functions expect. Let me know if you'd like to see anything changed here :-).

basveeling commented on miio/__init__.py:
```diff
@@ -42,6 +42,7 @@
 from miio.heater import Heater
 from miio.heater_miot import HeaterMiot
 from miio.huizuo import Huizuo, HuizuoLampFan, HuizuoLampHeater, HuizuoLampScene
+from miio.integrations.ihcooker.ihcooker import IHCooker
```
Shuffled most of the constants to recipe_profile to prevent circular imports.

basveeling commented on miio/integrations/ihcooker/tests/test_ihcooker.py:
```diff
@@ -0,0 +1,321 @@
+        )
+
+        res = profile_v2.parse(bytes.fromhex(recipe))
+        self.assertEqual(profile_v2.parse(profile_v2.build(res)), res)
```
Done! What is the reason for this?

rytilahti replied:
The use of testcase classes is a remnant from unittest's java roots. The preferred way nowadays is to avoid that class altogether and use individual test functions, or that's my understanding at least.

One of the reasons is that it's also cleaner and feels more idiomatic python, as you don't need to think about which [assert function](https://docs.python.org/3/library/unittest.html#assert-methods) should be used.

rytilahti reviewed (changes requested):
Could you move the directroy to be under `cooker/chunmi/` (just in case some other "cooker" devices come around), and rebase on top of the current master? Otherwise this is good to go and should be a part of the next release.

dmatora commented:
I have chunmi.ihcooker.v2
tried running `miiocli cooker --ip blablabla --token blablabla status` and got
```
WARNING:miio.device:Found an unsupported model 'chunmi.ihcooker.v2' for class 'Cooker'. If this is working for you, please open an issue at https://github.com/rytilahti/python-miio/
Error: {'code': -9999, 'message': 'user ack timeout'}
```

dmatora commented:
@syssi where can I read about extracting react plugin from android device?

trackhacs commented:
what to do to reactivate this?

thewh1teagle commented:
Hey everyone,

For a long time, I wanted to control my induction cooker with my phone without using the Xiaomi app. I also wanted to set the temperature myself instead of relying on pre-made recipes. So, I followed a pull request and created an Android app that supports version 2 of the cooker. Give it a try!

Thanks, and I hope this app helps make your life easier!

https://github.com/thewh1teagle/MiCook

thewh1teagle commented:
@basveeling 
Maybe you will be interested, and it will be great if you can take a look
I rewrote this PR in Dart and use it in a cross-platform Flutter app which controls the cooker. 
Here is the repository: https://github.com/thewh1teagle/MiCook.

tuxcrafter commented:
Any news on this merge for the cookers?
